src/mds/journal.cc

```diff
@@ -2505,36 +2509,29 @@ void ESlaveUpdate::replay(MDSRank *mds)
   case ESlaveUpdate::OP_PREPARE:
     dout(10) << "ESlaveUpdate.replay prepare " << reqid << " for mds." << master 
 	     << ": applying commit, saving rollback info" << dendl;
-    su = new MDSlaveUpdate(origop, rollback, segment->slave_updates);
+    su = new MDSlaveUpdate(origop, rollback);
     commit.replay(mds, segment, su);
-    mds->mdcache->add_uncommitted_slave_update(reqid, master, su);
+    mds->mdcache->add_uncommitted_slave(reqid, segment, master, su);
     break;
 
   case ESlaveUpdate::OP_COMMIT:
-    su = mds->mdcache->get_uncommitted_slave_update(reqid, master);
-    if (su) {
-      dout(10) << "ESlaveUpdate.replay commit " << reqid << " for mds." << master << dendl;
-      mds->mdcache->finish_uncommitted_slave_update(reqid, master);
-    } else {
-      dout(10) << "ESlaveUpdate.replay commit " << reqid << " for mds." << master 
-	       << ": ignoring, no previously saved prepare" << dendl;
-    }
+    dout(10) << "ESlaveUpdate.replay commit " << reqid << " for mds." << master << dendl;
+    mds->mdcache->finish_uncommitted_slave(reqid);
```

Comment:
ops, this seems buggy. `finish_uncommitted_slave` may be asserted. I'll  try to find a proper way.

Reply:
finish_uncommitted_slave() may find that  reqid is not in uncommitted_slaves. If that happens,  finish_uncommitted_slave() should just return.  we can add a parameter to finish_uncommitted_slave(), indicating if reqid not found is allowed

Reply:
I added `assert_exist` parameter with a default value being true. The false value should only be used in two situation:
1. doing replay, because PREPARE log may have been trimmed
2. sometimes in those commit/rollback callback functions, because they are used both in master-slave scene and non-slave scene, so we have to decide it with mdr->slave_update_journaled